### ISSUE TITLE
Insertion sort leads to a timeout on Z3

### ISSUE THREAD

**Issue by b-studios** [epfl-lara/stainless#1305]
Stainless is looking really great! I played around with a few examples, but couldn't verify insertion sort.

```scala
  def smallerThanAll(el: Int, list: List[Int]): Boolean = list match {
    case Nil() => true
    case Cons(head, tail) => el <= head && smallerThanAll(el, tail)
  }

  def sorted(list: List[Int]): Boolean = list match {
    case Nil() => true
    case Cons(head, tail) => smallerThanAll(head, tail) && sorted(tail)
  }

  def insertInto(el: Int, lst: List[Int]): List[Int] = {
    require(sorted(lst))
    lst match {
      case Nil() => List(el)
      case Cons(head, tail) => if (el <= head) Cons(el, lst) else Cons(head, insertInto(el, tail))
    }
  } ensuring { sorted }
```

It always leads to a timeout:
```
[error] -- Error: /private/tmp/designbycontract/verified/src/main/scala/verified.scala:118:98
[error] 118 |      case Cons(head, tail) => if (el <= head) Cons(el, lst) else Cons(head, insertInto(el, tail))
[error]     |                                                                  ^
[error]     |                                  VC 'postcondition' @ 118:67: TIMEOUT
[error]     |                                   => TIMEOUT
[warn]  - Result for 'postcondition' VC for insertInto @118:67:
[info]  - Checking cache: 'body assertion: match exhaustiveness' VC for insertInto @116:5...
[info] Cache miss: 'body assertion: match exhaustiveness' VC for insertInto @116:5...
[info]  - Now solving 'body assertion: match exhaustiveness' VC for insertInto @116:5...
[info] !sorted(lst) || true
```

I already tried to align the recursion structure of the sorted predicate to the recursion structure of `insertInto` (something that one always would try with theorem provers), but to no avail.

Do you have any hints on how I can proceed? BTW, is it possible to pass options (like timeout) when using the sbt plugin?

**Comment by b-studios** [comment]
Embarrassingly, I just noticed that sorting is in the tutorial: https://epfl-lara.github.io/stainless/tutorial.html

**Comment by mario-bucev** [comment]
Thank you for you interest in Stainless 😃 ! The documentation is a bit all over the place, we need to clean it up a bit.
As for the SBT plugin, it's quite unwieldy to use (as you may have noticed!) that's why we usually recommend going for the release https://github.com/epfl-lara/stainless/releases

**Comment by b-studios** [comment]
Hey @mario-bucev, thanks for your kind answer. I am considering to use stainless for teaching a software engineering course (to demonstrate that pre- and postconditions can be statically checked).

I thought an sbt plugin would be the easiest for the students to get started, especially if it would ship an instance of Z3 with it. This way, we would only provide them with a project template and sbt will do the rest.

BTW the tutorial and documentation is really good! Strangely, however, the example

```
  def rightUnitAppend[T](l1: List[T]): Boolean = {
    l1 ++ Nil() == l1
  }.holds
```
which is claimed to not be verifiable easily just goes through for me. Maybe stainless improved significantly?

**Comment by mario-bucev** [comment]
That's great to hear! The archive is "ready-to-use" and already contains a Z3 binary (v. 4.8.14). If you would like to use the SBT plugin, we'll gladly dust it off :p

Also thank you for reporting that `rightUnitAppend` verifies (I've tested it and verifies as well). We'll try to replace that example with something else.

**Comment by b-studios** [comment]
Some comments wrt to the release package:

- I had to rename `stainless.sh` to `stainless` and make it executable.
- It looks like on MacOS I am running into code-signing issues with the bundled z3 binary. (I know my way around this, but students might not :) )
- The shipped z3 binary for Mac OS is x86 -- maybe that will result in performance issues:

```
~/c/s/z3 file z3
z3: Mach-O 64-bit executable x86_64
```

There are also some other advantages of your sbt plugin:
- for me it immediately worked with Scala 3, just by choosing `3.0.2` as the `scalaVersion`.
- roundtrip time is faster than running the standalone executable

So from my side, your sbt pluging worked really well. It is just that `stainless` can't be fetched from the repositories anymore and I had to run `publishLocal`, which might be too much for students to ask (if we only demo stainless in a single lecture).
-
